internal/appsec/config/wafmanager.go

```diff
@@ -0,0 +1,206 @@
+		defaultLoaded bool
+		closed        bool
+		// mu is used to guard access to the [WAFManager] state.
+		mu sync.RWMutex
+	}
+
+	// WAFManagerUpdater is a helper obtained from [WAFManager.LockForUpdates] that is used to perform
+	// changes to the current configuration loaded in the [WAFManager]. The [WAFManagerUpdater.Unlock]
+	// function must be called once done.
+	WAFManagerUpdater struct {
+		manager *WAFManager
+	}
+)
+
+const defaultRulesPath = "ASM_DD/default"
+
+// NewWAFManager creates a new [WAFManager] with the provided [appsec.ObfuscatorConfig] and initial
+// rules (if any).
+func NewWAFManager(obfuscator appsec.ObfuscatorConfig, defaultRules any) (*WAFManager, error) {
+	builder, err := libddwaf.NewBuilder(obfuscator.KeyRegex, obfuscator.ValueRegex)
+	if err != nil {
+		return nil, err
+	}
+
+	rulesVersion := ""
+	if defaultRules != nil {
+		diag, err := builder.AddOrUpdateConfig(defaultRulesPath, defaultRules)
+		if err != nil {
+			builder.Close()
+			return nil, err
+		}
+		diag.EachFeature(func(name string, feature *libddwaf.Feature) {
+			if feature.Error != "" {
+				log.Error("%s", feature.Error, telemetry.WithTags([]string{"appsec_config_key:" + name, "log_type:local::diagnostic"}))
+			}
+			for msg, ids := range feature.Errors {
+				log.Error("%s: %q", msg, ids, telemetry.WithTags([]string{"appsec_config_key:" + name, "log_type:local::diagnostic"}))
+			}
+			for msg, ids := range feature.Warnings {
+				log.Warn("%s: %q", msg, ids, telemetry.WithTags([]string{"appsec_config_key:" + name, "log_type:local::diagnostic"}))
+			}
+		})
+		rulesVersion = diag.Version
+	}
+
+	mgr := &WAFManager{
+		builder:       builder,
+		initRules:     defaultRules,
+		rulesVersion:  rulesVersion,
+		defaultLoaded: defaultRules != nil,
+	}
+	// Attach a finalizer to close the builder when it is garbage collected, in case
+	// [WAFManager.Close] is not called explicitly by the user. The call to [libddwaf.Builder.Close]
+	// is safe to make multiple times.
+	runtime.SetFinalizer(mgr, func(m *WAFManager) { m.doClose(true) })
+
+	return mgr, nil
+}
+
+// LockForUpdates returns a [WAFManagerUpdater] that can be used to perform multiple updates to the
+// receiving [WAFManager] while preventing [WAFManager.NewHandle] from creating new handles with
+// partially updated configurations. The caller must call [WAFManagerUpdater.Unlock] once done. The
+// goroutine that called [WAFManager.LockForUpdates] must be careful not to call other methods on
+// the receiving [WAFManager] until [WAFManagerUpdater.Unlock] has been called, as these will result
+// in a deadlock.
+func (m *WAFManager) LockForUpdates() *WAFManagerUpdater {
+	m.mu.Lock()
+	return &WAFManagerUpdater{manager: m}
+}
+
+// Reset resets the WAF manager to its initial state.
+func (m *WAFManager) Reset() error {
+	bulk := m.LockForUpdates()
+	for _, path := range m.builder.ConfigPaths("") {
+		bulk.RemoveConfig(path)
+	}
+	err := bulk.RestoreDefaultConfig()
+	bulk.Unlock()
+	return err
```

Comment:
```suggestion
	bulk := m.LockForUpdates()
	defer bulk.Unlock()
	for _, path := range m.builder.ConfigPaths("") {
		bulk.RemoveConfig(path)
	}
	return bulk.RestoreDefaultConfig()
```


Other comments would make it so you don't have to apply this comment but we never know